common/metrics/otel_metric_provider.go

```diff
@@ -0,0 +1,142 @@
+		provider:    o,
+		excludeTags: configExcludeTags(cfg),
+	}
+}
+
+// WithTags creates a new MetricProvder with provided []Tag
+// Tags are merged with registered Tags from the source MetricsHandler
+func (omp *otelMetricsHandler) WithTags(tags ...Tag) MetricsHandler {
+	return &otelMetricsHandler{
+		provider:    omp.provider,
+		excludeTags: omp.excludeTags,
+		tags:        append(omp.tags, tags...),
+	}
+}
+
+// Counter obtains a counter for the given name and MetricOptions.
+func (omp *otelMetricsHandler) Counter(counter string) CounterMetric {
+	c, err := omp.provider.GetMeter().SyncInt64().Counter(counter)
+	if err != nil {
+		omp.l.Fatal("error getting metric", tag.NewStringTag("MetricName", counter), tag.Error(err))
+	}
+
+	return CounterMetricFunc(func(i int64, t ...Tag) {
+		c.Add(context.Background(), i, tagsToAttributes(omp.tags, t, omp.excludeTags)...)
+	})
+}
+
+// Gauge obtains a gauge for the given name and MetricOptions.
+func (omp *otelMetricsHandler) Gauge(gauge string) GaugeMetric {
+	c, err := omp.provider.GetMeter().SyncFloat64().UpDownCounter(gauge)
+	if err != nil {
+		omp.l.Fatal("error getting metric", tag.NewStringTag("MetricName", gauge), tag.Error(err))
+	}
+
+	return GaugeMetricFunc(func(i float64, t ...Tag) {
+		c.Add(context.Background(), i, tagsToAttributes(omp.tags, t, omp.excludeTags)...)
+	})
+}
+
+// Timer obtains a timer for the given name and MetricOptions.
+func (omp *otelMetricsHandler) Timer(timer string) TimerMetric {
+	c, err := omp.provider.GetMeter().SyncInt64().Histogram(timer, instrument.WithUnit(otelunit.Unit(Milliseconds)))
+	if err != nil {
+		omp.l.Fatal("error getting metric", tag.NewStringTag("MetricName", timer), tag.Error(err))
+	}
+
+	return TimerMetricFunc(func(i time.Duration, t ...Tag) {
+		c.Record(context.Background(), i.Nanoseconds(), tagsToAttributes(omp.tags, t, omp.excludeTags)...)
+	})
+}
+
+// Histogram obtains a histogram for the given name and MetricOptions.
+func (omp *otelMetricsHandler) Histogram(histogram string, unit MetricUnit) HistogramMetric {
+	c, err := omp.provider.GetMeter().SyncInt64().Histogram(histogram, instrument.WithUnit(otelunit.Unit(unit)))
+	if err != nil {
+		omp.l.Fatal("error getting metric", tag.NewStringTag("MetricName", histogram), tag.Error(err))
+	}
+
+	return CounterMetricFunc(func(i int64, t ...Tag) {
+		c.Record(context.Background(), i, tagsToAttributes(omp.tags, t, omp.excludeTags)...)
+	})
+}
+
+func (omp *otelMetricsHandler) Stop(l log.Logger) {
+	omp.provider.Stop(l)
+}
+
+// tagsToAttributes helper to merge registred tags and additional tags converting to attribute.KeyValue struct
+func tagsToAttributes(t1 []Tag, t2 []Tag, e excludeTags) []attribute.KeyValue {
+	var attrs []attribute.KeyValue
+
+	convert := func(tag Tag) attribute.KeyValue {
+		if vals, ok := e[tag.Key()]; ok {
+			if _, ok := vals[tag.Value()]; ok {
+				return attribute.String(tag.Key(), tagExcludedValue)
+			}
+		}
+
+		return attribute.String(tag.Key(), tag.Value())
```

Comment:
I think we have bug here. The second if should be !ok. See exclude tag: https://github.com/temporalio/temporal/blob/b7262fe3c79120d695750e093d09b5627e989fcb/common/metrics/config.go#L57-L60

Same apply to tally provider.
We should have some unit test for this one.

Reply:
Also note, the comment suggest the excluded tag will have value of `__disabled__`, but it actually has value of `_tag_excluded_`. Please update that as well.